enterprise/signals/utils.py

```diff
@@ -31,6 +32,73 @@
 logger = logging.getLogger(__name__)
 
 
+def simulate(pta, params, sparse_cholesky=True):
+    """Simulate residuals for all pulsars in `pta` by sampling all white-noise
+    and GP objects for parameters `params`. Requires GPs to have `combine=False`,
+    and will run faster with GP ECORR. If `pta` includes a `TimingModel`, that
+    should be created with a small `prior_variance`. This function can be used
+    with `utils.set_residuals` to replace residuals in a `Pulsar` object.
+    Note that any PTA built from that `Pulsar` may nevertheless cache residuals
+    internally, so it is safer to rebuild the PTA with the modified `Pulsar`."""
+
+    delays, ndiags, fmats, phis = (
+        pta.get_delay(params=params),
+        pta.get_ndiag(params=params),
+        pta.get_basis(params=params),
+        pta.get_phi(params=params),
+    )
+
+    gpresiduals = []
+    if pta._commonsignals:
+        if sparse_cholesky:
+            cf = cholesky(sps.csc_matrix(phis))
+            gp = np.zeros(phis.shape[0])
+            gp[cf.P()] = np.dot(cf.L().toarray(), np.random.randn(phis.shape[0]))
+        else:
+            gp = np.dot(sl.cholesky(phis, lower=True), np.random.randn(phis.shape[0]))
+
+        i = 0
+        for fmat in fmats:
+            j = i + fmat.shape[1]
+            gpresiduals.append(np.dot(fmat, gp[i:j]))
+            i = j
+
+        assert len(gp) == i
+    else:
+        for fmat, phi in zip(fmats, phis):
+            if phi is None:
+                gpresiduals.append(0)
+            elif phi.ndim == 1:
+                gpresiduals.append(np.dot(fmat, np.sqrt(phi) * np.random.randn(phi.shape[0])))
+            else:
+                raise NotImplementedError
+
+    whiteresiduals = []
+    for delay, ndiag in zip(delays, ndiags):
+        if ndiag is None:
+            whiteresiduals.append(0)
+        elif isinstance(ndiag, ShermanMorrison):
```

Comment:
Not a good idea to check for an instance of ShermanMorrison. When 'fastshermanmorrison' is used, this will be a different type. Instead, perhaps duck typing can be used, like:

`if all(hasattr(ndiag, attr) for attr in ['_nvec', '_jvec', '_slices']):`